File: lib/pricing-engine/conventions.ts
import { generateCouponDates } from "./cashflow";

export const COUPON_FREQUENCY = 2;
export const DAY_COUNT = "ACT/ACT";
export const SETTLEMENT_DAYS = 1;

export function getSettlementDate(tradeDate: Date): Date {
  const date = new Date(tradeDate.getTime());
  const day = date.getDay(); // 0 = Sunday, 6 = Saturday
  
  if (day === 5) { // Friday trade -> settles Monday (T+3)
    date.setDate(date.getDate() + 3);
  } else if (day === 6) { // Saturday trade -> settles Monday (T+2)
    date.setDate(date.getDate() + 2);
  } else if (day === 0) { // Sunday trade -> settles Tuesday (T+2)
    date.setDate(date.getDate() + 2);
  } else { // Mon-Thu trade -> settles next business day (T+1)
    date.setDate(date.getDate() + 1);
  }
  return date;
}

export function calculateAccruedInterest(
  settlementDate: Date,
  issueDate: Date,
  maturityDate: Date,
  couponRate: number,
  couponFrequency: number = 2,
  faceValue: number = 100.0
): number {
  if (settlementDate.getTime() <= issueDate.getTime()) {
    return 0.0;
  }
  if (settlementDate.getTime() >= maturityDate.getTime()) {
    return 0.0;
  }

  const dates = generateCouponDates(issueDate, maturityDate, couponFrequency);
  
  let activeIdx = -1;
  for (let i = 0; i < dates.length - 1; i++) {
    const periodStart = i === 0 ? issueDate : dates[i];
    const periodEnd = dates[i + 1];
    
    if (settlementDate.getTime() >= periodStart.getTime() && settlementDate.getTime() < periodEnd.getTime()) {
      activeIdx = i;
      break;
    }
  }

  if (activeIdx === -1) {
    return 0.0;
  }

  const tStart = dates[activeIdx];
  const tEnd = dates[activeIdx + 1];
  
  const periodDays = Math.round((tEnd.getTime() - tStart.getTime()) / (1000 * 60 * 60 * 24));
  if (periodDays === 0) {
    return 0.0;
  }

  const accrualStart = activeIdx === 0 ? issueDate : tStart;
  const accruedDays = Math.round((settlementDate.getTime() - accrualStart.getTime()) / (1000 * 60 * 60 * 24));
  
  const couponPayment = faceValue * (couponRate / 100.0) / couponFrequency;
  return couponPayment * (accruedDays / periodDays);
}

File: backend/quant_core/conventions.py
from datetime import date, timedelta
from typing import List

# Centralized G-Sec market conventions per Quant Methodology §7.2
COUPON_FREQUENCY = 2   # Semi-annual
DAY_COUNT = "ACT/ACT"  # Actual/Actual (ICMA)
SETTLEMENT_DAYS = 1    # T+1

def get_settlement_date(trade_date: date) -> date:
    """
    Returns the settlement date (T+1 business days) for a given trade date.
    Excludes weekends (Saturdays and Sundays).
    
    # Weekend-aware settlement. Ignored Indian national holidays since no calendar is loaded.
    # Upgrade path: Integrate a holiday library or load holiday dates from DB when scaling.
    """
    weekday = trade_date.weekday()  # 0 = Monday, 6 = Sunday
    if weekday == 4:    # Friday -> Monday
        return trade_date + timedelta(days=3)
    elif weekday == 5 or weekday == 6:  # Saturday/Sunday -> Tuesday
        return trade_date + timedelta(days=2)
    else:               # Monday-Thursday -> Next Day
        return trade_date + timedelta(days=1)

def calculate_accrued_interest(
    settlement_date: date,
    issue_date: date,
    maturity_date: date,
    coupon_rate: float,
    coupon_frequency: int = 2,
    face_value: float = 100.0
) -> float:
    """
    Calculates the accrued interest for a bond.
    Uses the Actual/Actual (ICMA) day count convention.
    """
    if settlement_date <= issue_date:
        return 0.0
        
    if settlement_date >= maturity_date:
        return 0.0

    # Import schedule builder here to avoid circular imports
    from quant_core.cashflow import generate_coupon_dates
    
    dates = generate_coupon_dates(issue_date, maturity_date, coupon_frequency)
    
    # Find the active coupon period containing settlement_date
    # dates is [t_0, t_1, t_2, ..., t_N]
    active_idx = -1
    for i in range(len(dates) - 1):
        period_start = issue_date if i == 0 else dates[i]
        period_end = dates[i+1]
        
        if period_start <= settlement_date < period_end:
            active_idx = i
            break
            
    if active_idx == -1:
        return 0.0

    t_start = dates[active_idx]      # Theoretical or actual preceding coupon date (t_i)
    t_end = dates[active_idx + 1]    # Next coupon date (t_{i+1})
    
    period_days = (t_end - t_start).days
    if period_days == 0:
        return 0.0
        
    # Accrued days starts from actual start of interest accumulation
    accrual_start = issue_date if active_idx == 0 else t_start
    accrued_days = (settlement_date - accrual_start).days
    
    coupon_payment = face_value * (coupon_rate / 100.0) / coupon_frequency
    accrued_interest = coupon_payment * (accrued_days / period_days)
    
    return accrued_interest
